src/Compilers/Helpers/CoreClrAnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Loader;
+using Roslyn.Utilities;
+
+using static Microsoft.CodeAnalysis.AnalyzerAssemblyLoadUtils;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// Core CLR compatible wrapper for loading analyzers.
+    internal class CoreClrAnalyzerAssemblyLoader : AssemblyLoadContext, IAnalyzerAssemblyLoader
+    {
+        private readonly Dictionary<string, Assembly> _pathsToAssemblies = new Dictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, Assembly> _namesToAssemblies = new Dictionary<string, Assembly>();
+        private readonly List<string> _dependencyPaths = new List<string>();
+        private readonly object _guard = new object();
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CoreClrAnalyzerAssemblyLoader" />,
+        /// sets that instance to be the default <see cref="AssemblyLoadContext" />,
+        /// and returns that instance. Throws if the Default is already set or the
+        /// binding model is already locked.
+        /// </summary>
+        public static CoreClrAnalyzerAssemblyLoader CreateAndSetDefault()
+        {
+            var assemblyLoader = new CoreClrAnalyzerAssemblyLoader();
+            InitializeDefaultContext(assemblyLoader);
+            return assemblyLoader;
+        }
+
+        public void AddDependencyLocation(string fullPath)
+        {
+            if (fullPath == null)
+            {
+                throw new ArgumentNullException(nameof(fullPath));
+            }
+
+            lock (_guard)
+            {
+                if (!_dependencyPaths.Contains(fullPath, StringComparer.OrdinalIgnoreCase))
+                {
+                    _dependencyPaths.Add(fullPath);
```

Comment:
Should skip the `Contains` step here.  It is duplicated in the `Add` code path and causes double lookups.